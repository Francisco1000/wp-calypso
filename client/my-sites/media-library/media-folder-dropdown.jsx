/** @format */

/**
 * External dependencies
 */
import React, { Component } from 'react';
import { bindAll } from 'lodash';
import classNames from 'classnames';
import PropTypes from 'prop-types';
import { localize } from 'i18n-calypso';
/**
 * Internal dependencies
 */
import SelectDropdown from 'components/select-dropdown';

export class MediaFolderDropdown extends Component {
	static propTypes = {
		initialSelected: PropTypes.oneOfType( [ PropTypes.string, PropTypes.number ] ),
		folders: PropTypes.arrayOf(
			PropTypes.shape( {
				ID: PropTypes.oneOfType( [ PropTypes.string, PropTypes.number ] ).isRequired,
				URL: PropTypes.string,
				children: PropTypes.number.isRequired,
				date: PropTypes.string,
				file: PropTypes.string,
				name: PropTypes.string.isRequired,
				path: PropTypes.string.isRequired,
				thumbnails: PropTypes.object,
				type: PropTypes.string,
			} )
		),
		folder: PropTypes.string,
		onFolderChange: PropTypes.func,
		disabled: PropTypes.bool,
		compact: PropTypes.bool,
		defaultOption: PropTypes.object,
	};

	static defaultProps = {
		initialSelected: '/',
		folders: [],
		compact: true,
	};

	constructor( props ) {
		super( props );

		bindAll( this, [ 'handleOnSelect' ] );
	}

	handleOnSelect( option ) {
		if ( this.props.disabled ) return;

		const folder = option.value;

		this.props.onFolderChange( folder );
	}

	getDropDownOptions( folderData ) {
		const separator = null;

		return [
			{
				value: '/',
				label: this.props.translate( 'All Albums' ),
			},
			separator,
		].concat(
			folderData.map( folder => {
				return {
					value: '' + folder.ID, // convert to string if number
					label: folder.name,
				};
			} )
		);
	}

	render() {
		const rootClassNames = classNames( this.props.className, {
			'media-library__folder-dropdown': true,
		} );

		const folderOptions = this.getDropDownOptions( this.props.folders );

		// No need to show folders if we only have the default option + seperator
		if ( folderOptions.length <= 2 ) return null;

		return (
			<div className={ rootClassNames }>
				<SelectDropdown
					initialSelected={ this.props.folder } // convert to string if number
					disabled={ this.props.disabled }
					compact={ this.props.compact }
					onSelect={ this.handleOnSelect }
					options={ folderOptions }
				/>
			</div>
		);
	}
}

export default localize( MediaFolderDropdown );
