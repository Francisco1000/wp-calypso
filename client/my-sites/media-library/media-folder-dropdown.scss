.media-library__folder-dropdown {
}

.media-library__folder-dropdown-label {
	@include hide-content-accessibly();
}

.media-library__folder-dropdown-field {
	padding-top: 5px;
	padding-bottom: 5px;
	color: $gray-text-min;
	font-size: 12px;
	width: 140px;
	min-width: initial;
	font-weight: 500;

	&[disabled] {
		color: $gray-lighten-30;
		border-color: $gray-lighten-30;
		cursor: default;
	}
}
