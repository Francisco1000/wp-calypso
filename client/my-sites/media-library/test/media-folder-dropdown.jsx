/**
 * @format
 * @jest-environment jsdom
 */

/**
 * External dependencies
 */
import React from 'react';
import { shallow } from 'enzyme';

/**
 * Internal dependencies
 */
import { folders } from './fixtures';
import { MediaFolderDropdown } from '../media-folder-dropdown';
import SelectDropdown from 'components/select-dropdown';

const noop = rtn => rtn;

describe( 'MediaFolderDropdown', () => {
	let wrapper;
	let FIXTURE_FOLDERS_DATA;
	let FIXTURE_FOLDERS_OPTIONS;

	beforeEach( () => {
		if ( wrapper ) {
			wrapper.unmount();
		}

		FIXTURE_FOLDERS_DATA = folders;

		// Convert raw API folder data to the label/value key pairs
		// expected
		FIXTURE_FOLDERS_OPTIONS = FIXTURE_FOLDERS_DATA.map( folder => {
			return {
				value: folder.ID.toString(),
				label: folder.title,
			};
		} );
	} );

	test( 'default rendering', () => {
		wrapper = shallow(
			<MediaFolderDropdown translate={ noop } folders={ FIXTURE_FOLDERS_DATA } />
		);
		expect( wrapper.find( '.media-library__folder-dropdown' ).exists() ).toBe( true );
		expect( wrapper.find( SelectDropdown ).exists() ).toBe( true );
	} );

	test( 'does not render if no folders are provided', () => {
		wrapper = shallow( <MediaFolderDropdown translate={ noop } /> );
		expect( wrapper.find( '.media-library__folder-dropdown' ) ).toHaveLength( 0 );
	} );

	describe( 'initial folder selection', () => {
		test( 'initial folder is correctly defaulted to all folders', () => {
			wrapper = shallow(
				<MediaFolderDropdown translate={ noop } folders={ FIXTURE_FOLDERS_DATA } />
			);
			expect( wrapper.find( SelectDropdown ).props().initialSelected ).toBe( 'all' );
		} );

		test( 'initial folder can be overidden via props', () => {
			const expected = FIXTURE_FOLDERS_DATA[ 0 ].ID.toString();

			wrapper = shallow(
				<MediaFolderDropdown
					translate={ noop }
					folders={ FIXTURE_FOLDERS_DATA }
					initialSelected={ expected }
				/>
			);

			expect( wrapper.find( SelectDropdown ).props().initialSelected ).toBe( expected );
		} );
	} );

	test( 'folder options correctly merge with default items', () => {
		const expected = [
			{
				value: 'all',
				label: 'All Photos',
			},
			null, // seperator
		].concat( FIXTURE_FOLDERS_OPTIONS );

		wrapper = shallow(
			<MediaFolderDropdown translate={ noop } folders={ FIXTURE_FOLDERS_DATA } />
		);

		expect( wrapper.find( SelectDropdown ).props().options ).toEqual( expected );
	} );

	test( 'selected folder option value is correctly passed as argument to onFolderChange prop', () => {
		const spy = jest.fn();
		const selectedOption = FIXTURE_FOLDERS_DATA[ 0 ];
		const expected = selectedOption.value;

		wrapper = shallow(
			<MediaFolderDropdown
				translate={ noop }
				folders={ FIXTURE_FOLDERS_DATA }
				onFolderChange={ spy }
			/>
		);
		const cb = wrapper.find( SelectDropdown ).props().onSelect;

		// Call the spy
		cb( selectedOption );

		expect( spy ).toBeCalledWith( expected );
	} );

	test( 'correctly forwards disabled prop to SelectDropdown', () => {
		wrapper = shallow(
			<MediaFolderDropdown translate={ noop } folders={ FIXTURE_FOLDERS_DATA } disabled={ true } />
		);

		expect( wrapper.find( SelectDropdown ).props().disabled ).toBe( true );

		wrapper.setProps( { disabled: false } );

		expect( wrapper.find( SelectDropdown ).props().disabled ).toBe( false );
	} );
} );
