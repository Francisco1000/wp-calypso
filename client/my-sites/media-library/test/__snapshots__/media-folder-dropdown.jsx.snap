// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MediaFolderDropdown default rendering 1`] = `
<div
  className="media-library__folder-dropdown"
>
  <SelectDropdown
    compact={true}
    initialSelected="__all__"
    onSelect={[Function]}
    onToggle={[Function]}
    options={
      Array [
        Object {
          "label": "All Photos",
          "value": "__all__",
        },
        null,
        Object {
          "count": 9,
          "label": "Album 1",
          "value": "album-1",
        },
        Object {
          "count": 29,
          "label": "Album 2",
          "value": "album-2",
        },
        Object {
          "count": 52,
          "label": "Album 3",
          "value": "album-3",
        },
        Object {
          "count": 18,
          "label": "Album 4",
          "value": "album-4",
        },
      ]
    }
    style={Object {}}
  />
</div>
`;
